### PR TITLE
feat: virtual column support

### PR DESCRIPTION
**Pull request by xMase**
This feature was born from the need to make filters on the virtual columns of an entity. 
This code is a draft and for now it works only on virtual columns inside a relationships and with a limited set of filter operation. 

Compromise to make this:
1. Replaced Take-Skip with Limit-Offset to avoid syntax error in query generation (this also fix the problem of very complex query during the pagination phase) 
2. It was chosen to continue to use the Where operator instead of Having to simplify the code and make all operations compatible (this implies the execution of the sub-queries 2 times, this should hit the internal cache of the database and not affect performance)

Comments and proposals are very welcome.

### CODE REVIEW

**Comment by ppetzold** [comment]
@xMase PR is still in draft, or is it ready for review?

**Comment by xMase** [comment]
it's still a draft, we need to develop the virtual columns on the main entities and fix the search. If you have time to have a look at the code and make some suggestions, i'd appreciate it.

File: src/paginate.ts [hunk 1]
```diff
@@ -301,16 +299,16 @@ function generatePredicateCondition(
     ) as WherePredicateOperator
 }
 
-function addWhereConditionWithRelationVirtualEmbedded(
+function addWhereCondition(
     qb: SelectQueryBuilder<unknown>,
     column: string,
+    propiertes: ColumnProperties,
```
Comment (FrancYescO) [inline src/paginate.ts:310]: typo

**Comment by xMase** [comment]
@ppetzold  i think its ready to be merged
@jakobwgnr can you check if the issue #446 persist?

**Comment by ppetzold** [comment]
> @ppetzold i think its ready to be merged @jakobwgnr can you check if the issue #446 persist?

good stuff. will have some time to review next week

**Comment by jakobwgnr** [comment]
Hi @ppetzold, @xMase - Nice work.
Regarding #446 mentioned by @xMase we have to take care here. Currently this PR would revert the changes done there.

**Comment by xMase** [comment]
do you have tested it? can you give a project example to test it?

File: src/paginate.spec.ts [hunk 2]
```diff
@@ -1361,26 +1406,6 @@ describe('paginate', () => {
         expect(result.links.current).toBe('?page=1&limit=20&sortBy=id:ASC&filter.age=$not:$null')
     })
 
-    it('should return result based on not null query', async () => {
```
Comment (ppetzold) [inline src/paginate.spec.ts:1364]: why has this case been removed?

File: src/paginate.ts [hunk 3]
```diff
@@ -175,6 +178,144 @@ export const DEFAULT_MAX_LIMIT = 100
...
+    alias: string,
+    isVirtualProperty = false
+): WherePredicateOperator {
+    return qb['getWherePredicateCondition'](
+        isVirtualProperty ? column : alias,
+        filter[column]
+    ) as WherePredicateOperator
+}
+
+function addWhereCondition(qb: SelectQueryBuilder<unknown>, column: string, filter: Filter) {
```
Comment (ppetzold) [inline src/paginate.ts:307]: can we scope the naming here a bit more?

e.g. `addFilterWhereCondition`

File: src/paginate.ts [hunk 4]
```diff
@@ -175,6 +178,144 @@ export const DEFAULT_MAX_LIMIT = 100
...
+            return `${alias}_${properties.propertyPath}.${properties.propertyName}` // include embeded property and relation property
+        }
+    } else if (isVirtualProperty) {
+        return query ? `(${query(`${alias}`)})` : `${alias}_${properties.propertyName}`
+    } else {
+        return `${alias}.${properties.propertyName}` //
+    }
+}
+
+function generatePredicateCondition(
```
Comment (ppetzold) [inline src/paginate.ts:294]: same, alternatively let's move all filter helper functions already to it's own file

./paginate/filter.ts

then it's okay to keep the broader names

File: src/paginate.ts [hunk 5]
```diff
@@ -175,6 +178,144 @@ export const DEFAULT_MAX_LIMIT = 100
...
+        condition['condition']['parameters'] = condition_params
+    } else {
+        condition.parameters = condition_params
+    }
+
+    return params
+}
+
+// This function is used to fix the column alias when using relation, embedded or virtual properties
+function fixColumnAlias(
```
Comment (ppetzold) [inline src/paginate.ts:272]: same, alternatively let's move all filter helper functions already to it's own file

./paginate/filter.ts

then it's okay to keep the broader names

File: src/paginate.ts [hunk 6]
```diff
@@ -175,6 +178,144 @@ export const DEFAULT_MAX_LIMIT = 100
...
+function checkIsRelation(qb: SelectQueryBuilder<unknown>, propertyPath: string): boolean {
+    if (!qb || !propertyPath) {
+        return false
+    }
+    return !!qb?.expressionMap?.mainAlias?.metadata?.hasRelationWithPropertyPath(propertyPath)
+}
+
+// This function is used to fix the query parameters when using relation, embeded or virtual properties
+// It will replace the column name with the alias name and return the new parameters
+function fixQueryParam(
```
Comment (ppetzold) [inline src/paginate.ts:218]: same

File: src/paginate.ts [hunk 7]
```diff
@@ -175,6 +178,144 @@ export const DEFAULT_MAX_LIMIT = 100
...
+    const propertyPath = column.split('.')
+    return propertyPath.length > 1
+        ? {
+              propertyPath: propertyPath[0],
+              propertyName: propertyPath.slice(1).join('.'), // the join is in case of an embedded entity
+          }
+        : { propertyName: propertyPath[0] }
+}
+
+function extractVirtualProperty(
```
Comment (ppetzold) [inline src/paginate.ts:193]: same, alternatively let's move all filter helper functions already to it's own file

./paginate/filter.ts

then it's okay to keep the broader names

File: src/paginate.ts [hunk 8]
```diff
@@ -175,6 +178,144 @@ export const DEFAULT_MAX_LIMIT = 100
 export const DEFAULT_LIMIT = 20
 export const NO_PAGINATION = 0
 
+type ColumnProperties = { propertyPath?: string; propertyName: string }
+
+function getPropertiesByColumnName(column: string): ColumnProperties {
```
Comment (ppetzold) [inline src/paginate.ts:183]: same, alternatively let's move all filter helper functions already to it's own file

./paginate/filter.ts

then it's okay to keep the broader names

**Review by ppetzold — CHANGES_REQUESTED** [review]

File: src/paginate.spec.ts (hunk 2 above)
Comment (FrancYescO) [reply src/paginate.spec.ts:1364]: duplicate of the upper one... or am i wrong ?

Comment (ppetzold) [reply src/paginate.spec.ts:1364]: seems right, good catch then :)

**Comment by xMase** [comment]
some functions are not only used in filtering but also in search and sorting, move them to a specific file (ex. filters) or specifically renaming them would compromise their meaning. the affected functions has been moved to the helper file.

File: src/paginate.ts [hunk 9]
```diff
@@ -297,25 +314,19 @@ export async function paginate<T extends ObjectLiteral>(
...
+                    const { isVirtualProperty, query: virtualQuery } = extractVirtualProperty(qb, property)
+                    const isRelation = checkIsRelation(qb, property.propertyPath)
+
+                    const alias = fixColumnAlias(property, qb.alias, isRelation, isVirtualProperty, virtualQuery)
+                    const condition: WherePredicateOperator = {
+                        operator: 'ilike',
+                        parameters: [alias, `:${column}`],
                     }
+
+                    qb.orWhere(qb['createWhereConditionExpression'](condition), {
```
Comment (ppetzold) [inline src/paginate.ts:327]: just checkin this line here:

https://github.com/typeorm/typeorm/blob/defb409f5650fed0b7a4ff2933208282a45572fb/src/query-builder/QueryBuilder.ts#L1050

seems like we can avoid regression by adding ::text cast to `condition.parameters[0]` for `postgres` and `cockroachdb`.

```typescript
if (
    driver.options.type === "postgres" ||
    driver.options.type === "cockroachdb"
) {
    condition.parameters[0] += '::text'
}
```

wdyt?

sad we don't have coverage to test for those drivers..

**Review by ppetzold — CHANGES_REQUESTED** [review]

File: src/paginate.ts (hunk 9 above)
Comment (xMase) [reply src/paginate.ts:331]: yes without a test we cannot check it deep. But for now and without any additional test by the involved parts lgtm.

**Review by ppetzold — APPROVED** [review]

**Comment by ppetzold** [comment]
@jakobwgnr would appreciate some feedback if this PR introduces a regression to your recent changes.

on first sight, it seems to be about the equivalent:

https://github.com/ppetzold/nestjs-paginate/pull/434#discussion_r1090462654

**Comment by ppetzold** [comment]
@jakobwgnr just following up here, did you have a chance to check the new release?

**Comment by jakobwgnr** [comment]
@ppetzold currently on vacation… can follow up earliest on monday. It‘s on my todo list!

**Comment by jakobwgnr** [comment]
Hey @ppetzold, @xMase 
Unfortunately my test was not successful

Please see below the error. The problem is that the parameters are not covered by " ".

Will open an issue for this

WHERE (\"user\".\"deleted\" = $1) AND (user.firstName::text ILIKE $2 OR user.lastName::text ILIKE $3 OR user.email::text ILIKE $4 OR user.role::text ILIKE $5) ORDER BY \"user\".\"lastName\" ASC LIMIT 20","parameters":[false,"%SUPERADMIN%","%SUPERADMIN%","%SUPERADMIN%","%SUPERADMIN%"],"driverError":{"length":92,"name":"error","severity":"ERROR","code":"42601","position":"6513","file":"scan.l","line":"1180","routine":"scanner_yyerror"},"length":92,"severity":"ERROR","code":"42601","position":"6513","file":"scan.l","line":"1180","routine":"scanner_yyerror"}

**Pull request merged**